pyiron_base/interfaces/lockable.py

```diff
@@ -0,0 +1,347 @@
+    ...     return [k for k, v in self.items() if isinstance(v, LockGroupDict)]
+    ...
+    ...   def _list_nodes(self):
+    ...     return [k for k, v in self.items() if not isinstance(v, LockGroupDict)]
+
+    >>> d = LockGroupDict(a=dict(c=1, d=2), b=LockGroupDict(c=1, d=2))
+    >>> d.lock()
+
+    Since the first item is a plain dict, it can still be mutated.
+
+    >>> type(d['a'])
+    <class 'dict'>
+    >>> d['a']['c'] = 23
+    >>> d['a']['c']
+    23
+
+    Where as the second will be locked from now on
+
+    >>> type(d['b'])
+    <class 'lockable.LockGroupDict'>
+    >>> d['b']['c'] = 23
+    Traceback (most recent call last):
+        ...
+    lockable.Locked: Object is currently locked!  Use unlocked() if you know what you are doing.
+    >>> d['b']['c']
+    1
+
+    but we can unlock it as usual
+
+    >>> with d.unlocked() as dopen:
+    ...   dopen['b']['d'] = 23
+    >>> d['b']['d']
+    23
+
+    To use this class with properties, simply decorate the setter
+
+    >>> class MyLock(Lockable):
+    ...   def __init__(self, foo):
+    ...     super().__init__()
+    ...     self._foo = foo
+    ...   @property
+    ...   def foo(self):
+    ...     return self._foo
+    ...   @foo.setter
+    ...   @sentinel
+    ...   def foo(self, value):
+    ...     self._foo = value
+    >>> ml = MyLock(42)
+    >>> ml.foo
+    42
+    >>> ml.foo = 23
+    >>> ml.lock()
+    >>> ml.foo = 42
+    Traceback (most recent call last):
+        ...
+    lockable.Locked: Object is currently locked!  Use unlocked() if you know what you are doing.
+
+    It's possible to change the errors raised into a warning and allow
+    modification by passing `lock_method` to :meth:`~.Lockable.__init__` or
+    `method` to :meth:`~.lock`.
+
+    >>> mw = LockList(lock_method="warning")
+    >>> mw.append(0)
+    >>> mw.lock()
+    >>> mw[0] = 1 # will print the warning
+    >>> mw[0]
+    1
+
+    >>> mw = LockList()
+    >>> mw.append(0)
+    >>> mw.lock(method='warning')
+    >>> mw[0] = 1 # will print the warning
+    >>> mw[0]
+    1
+
+    """
+
+    def __init__(self, *args, lock_method: str = "error", **kwargs):
+        object.__setattr__(self, "_read_only", False)
+        object.__setattr__(self, "_lock_method", lock_method)
+        super().__init__(*args, **kwargs)
+
+    @property
+    def read_only(self) -> bool:
+        """
+        bool: False if the object can currently be written to
+
+        Setting this value will trigger :meth:`._on_lock` and :meth:`._on_unlock` if it changes.
+        """
+        return object.__getattribute__(self, "_read_only")
+
+    @read_only.setter
+    def read_only(self, value: bool):
+        changed = self._read_only != value
+        if changed:
+            self._read_only = value
+            if value:
+                self._on_lock()
+            else:
+                self._on_unlock()
+
+    def _on_lock(self):
+        for it in _iterate_lockable_subs(self):
+            it.lock()
```

Comment:
```suggestion
        pass
```

maybe (see main review comment)